lightwood/analysis/nn_conf/temp_scale.py

```diff
@@ -0,0 +1,100 @@
+from copy import deepcopy
+from typing import Dict, Tuple
+from types import SimpleNamespace
+
+import torch
+import pandas as pd
+from torch import nn, optim
+from sklearn.preprocessing import OrdinalEncoder
+
+from lightwood.helpers.log import log
+from lightwood.analysis.base import BaseAnalysisBlock
+
+
+class TempScaler(BaseAnalysisBlock):
+    """
+    Original reference (MIT Licensed): https://github.com/gpleiss/temperature_scaling
+    NB: Output of the neural network should be the classification logits, NOT the softmax (or log softmax)! TODO
+    """
+    def __init__(self):
+        super().__init__()
+        self.temperature = nn.Parameter(torch.ones(1))
+        self.ordenc = OrdinalEncoder()
+        self._softmax = torch.nn.Softmax(dim=1)
+        self.active = False
+
+    def temperature_scale(self, logits):
+        temperature = self.temperature.unsqueeze(1).expand(logits.size(0), logits.size(1))
+        return logits / temperature
+
+    def softmax(self, logits):
+        return self._softmax(self.temperature_scale(logits))
+
+    def analyze(self, info: Dict[str, object], **kwargs) -> Dict[str, object]:
+        """
+        Tune and set the temperature of a neural model optimizing NLL using validation set logits.
+        """
+        ns = SimpleNamespace(**kwargs)
+
+        if ns.predictor.mixers[ns.predictor.indexes_by_accuracy[0]].supports_proba:
+            self.n_cls = len(ns.stats_info.train_observed_classes)
+            nll_criterion = nn.CrossEntropyLoss()
+            self.ordenc.fit([[val] for val in ns.stats_info.train_observed_classes])
+
+            # collect logits and labels for the validation set
+            logits_list = []
+            labels_list = []
+            with torch.no_grad():
+                prob_cols = [col for col in ns.normal_predictions.columns
+                             if '__mdb_proba' in col and
+                             '__mdb_unknown_cat' not in col]
+                if not prob_cols:
+                    return info  # early stop if no proba info is available
```

Comment:
I think we ought to instead *not* automatically add this module via json-ai when `predict_proba` is not set for at least 1 mixer.

If the user adds it manually that's fine, it's on them to have understanding that this is needed, more confusing if they add it and it doesn't run.

Reply:
I agree, and in fact that's what happens right now, as the default blocks in `JsonAI` remain unchanged.